First commits for an Ansible role to install and configure `systemd-resolved`

jsf9k opened the pull request:
## 🗣 Description ##

This PR creates an Ansible role to install and configure `systemd-resolved`.

## 💭 Motivation and context ##

This PR contributes to the resolution of cisagov/cool-system-internal#140 since, once this role is applied to our COOL AMIs, they will use the `systemd-resolved` stub DNS resolver.

## 🧪 Testing ##

All automated tests pass.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

jsf9k committed:
Update README boilerplate for this Ansible role

jsf9k committed:
Update role metadata for this Ansible role

jsf9k committed:
Update tasks and Molecule tests to add/test functionality

jsf9k committed:
Remove support for Debian Buster and Bullseye as well as Ubuntu Focal and Jammy

These platforms do not provide systemd-resolved.

jsf9k committed:
Add code to unmount /etc/resolv.conf in Molecule prepare stage

Docker bind mounts a file from the host to /etc/resolv.conf.  This is
inconvenient for us, since we need to create a symlink at
/etc/resolv.conf.  At the same time, we don't want to break DNS.  The
playbook being imported contains a workaround for this situation.

jsf9k committed:
Enable and start systemd-resolved sooner

We must start the service to populate the files in
/run/systemd/resolve/, before we can create the /etc/resolv.conf.

jsf9k committed:
Comment out call to Service.exists

This functionality from pytest-testinfra is currently broken.  See
pytest-dev/pytest-testinfra#757 for more details.  Once
pytest-dev/pytest-testinfra#754 has been merged and a new release of
pytest-testinfra is created the Service.exists line can be restored.

jsf9k committed:
Preserve user, group, and mode when copying around resolv.conf

jsf9k committed:
Do not change the owner, group, or mode of the target when creating the symlink

jsf9k committed:
Use a different symlink target for AmazonLinux2023

/run/systemd/resolve/stub-resolv.conf is a symlink to
/run/systemd/resolve/resolv.conf in AL2023, so in this case the
/etc/resolv.conf symlink resolves to the former not the latter.

jsf9k committed:
Add ansible as a dependency for the ansible-lint pre-commit hook

For some reason ansible-lint does not know about the existence of
ansible.posix.mount unless ansible itself is added as an extra
dependency.  I believe this is because ansible is not installed when
ansible-lint is installed.

mcdonnnj commented on .pre-commit-config.yaml:
```diff
@@ -157,6 +157,8 @@ repos:
     rev: v24.2.0
     hooks:
       - id: ansible-lint
+        additional_dependencies:
```
I think we should modify this to use the same `ansible` and `ansible-core` versions as we have pinned in `requirements-test.txt`. This might be worth kicking back up to [cisagov/skeleton-generic](https://github.com/cisagov/skeleton-generic) and making the default for that part of the `pre-commit` configuration. ~~We should also add a comment explaining why we want to override the hook's dependencies.~~ Update for the comment: per the [documentation](https://github.com/ansible/ansible-lint/blob/main/src/ansiblelint/rules/syntax_check.md#syntax-checkunknown-module) and [pre-commit hook configuration](https://github.com/ansible/ansible-lint/blob/ad0157eb38059b02d57458504340209f221e3189/.pre-commit-hooks.yaml#L14-L19), if it's not in `ansible-core` it won't be found.

jsf9k replied:
Please see commit [a68a9b5](https://github.com/cisagov/ansible-role-systemd-resolved/pull/1/commits/a68a9b58160c47c31eff02adef093302fbb43a67).

I think the PR you mentioned should perhaps be in [cisagov/skeleton-ansible-role](https://github.com/cisagov/skeleton-ansible-role) since it is there that we first pin `ansible` and `ansible-core`.  What do you think?

mcdonnnj replied:
Well I figured [cisagov/skeleton-generic](https://github.com/cisagov/skeleton-generic) is appropriate for pinning these for `ansible-lint` so that any project using Ansible will by default respect the pins used by our Ansible roles (this is with [cisagov/cyhy_amis](https://github.com/cisagov/cyhy_amis) in mind).

jsf9k replied:
Please see cisagov/skeleton-generic#180.

mcdonnnj replied:
As a quick closer question do we want to add `ansible` in the same way but commented out with an explanation about needing to enable it if you run into issues with built-in modules not being found (like here)?

jsf9k replied:
Personally I prefer to add it as needed, just to avoid another dependency that needs to be kept in sync across repos.

mcdonnnj replied:
How about just a comment that mentions the possible need if it happens in another downstream repository? That way there's no dependency to maintain but the knowledge of fixing that ansible-lint failure isn't lost.

jsf9k replied:
Please see commit cisagov/skeleton-generic@8e57244cbe10b3a3f28276c32395a0228f1d773f (part of cisagov/skeleton-generic#180).

mcdonnnj replied:
Appreciated!

dav3r commented on .pre-commit-config.yaml:
```diff
@@ -157,6 +157,8 @@ repos:
     hooks:
       - id: ansible-lint
+        additional_dependencies:
+          - ansible
```
Please add a comment here explaining why we need to include ansible as a dependency.

jsf9k replied:
Please see commit ab70d7f.

dav3r replied:
Excellent, thanks.

dav3r commented on molecule/default/tests/test_default.py:
```diff
@@ -4,15 +4,48 @@
+def test_services(host):
+    """Verify that the expected services are present."""
+    s = host.service("systemd-resolved")
+    # This assertion currently fails because of
+    # pytest-dev/pytest-testinfra#757.  Once
+    # pytest-dev/pytest-testinfra#754 has been merged and a new
+    # release is created the following line can be uncommented.
+    # assert s.exists, "systemd-resolved service does not exist."
```
This sounds like it should be a TODO with an associated issue.

jsf9k replied:
Please see commit 1255c63.

dav3r replied:
Thanks!

dav3r reviewed (approved):
Good stuff! 🚀 
I noted naught but two small thangs.

jsf9k committed:
Add a comment

The comment explains why ansible must be added as an additional
dependency for the ansible-lint linter.

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

jsf9k committed:
Create and reference issue for uncommenting of assertion in test code

Co-authored-by: dav3r <david.redmin@gwe.cisa.dhs.gov>

mcdonnnj commented on meta/requirements.yml:
I know it's kind of cursed but please unwind _all_ unintended changes from https://github.com/cisagov/ansible-role-systemd-resolved/commit/4b9e8d5e8fc326cad9acdb8d6680a617c2e73c36.

mcdonnnj replied:
Please see https://github.com/cisagov/gh-skeleton/issues/33 to track that undesired behavior.

jsf9k replied:
I performed some force-push chicanery on the `develop` branch to get rid of the undesired changes.  I essentially replaced the commit you mentioned with commit cf9d35d02d69d7ed0ae0187a578a0342d17e5bce.

jsf9k committed:
Verify the systemd-resolved stub resolver is being used by default

Note that this entails installing dnsutils in the Molecule prepare
stage so that dig is available when the Molecule tests are run.

mcdonnnj commented on tasks/main.yml:
```diff
@@ -1,2 +1,31 @@
+
+- name: Create /etc/resolv.conf symlink
+  ansible.builtin.file:
+    # Do not apply any group, owner, or mode changes to src.
```
Since we are not setting those parameters I am unclear on what this comment is implying happens based on the [documentation](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/file_module.html#parameter-follow).

jsf9k replied:
Please see commit [32e1a61](https://github.com/cisagov/ansible-role-systemd-resolved/pull/1/commits/32e1a617488785d161822bbeb2be96db8080c12e).

mcdonnnj replied:
Ahh thanks. My reading of the documentation definitely does not reflect your observed behavior.

mcdonnnj commented on molecule/default/unmount.yml:
```diff
@@ -0,0 +1,33 @@
+  become: true
+  become_method: ansible.builtin.sudo
+  tasks:
+    - name: >-
```
Would you please separate each of these tasks such as we do in `tasks/main.yml`? I just find the lack of separation unpleasant when reading through a list of tasks.

jsf9k replied:
Please see commit [9a51c3a](https://github.com/cisagov/ansible-role-systemd-resolved/pull/1/commits/9a51c3aa334a8d252e6ffc4349a2f547c4c87bd6).

mcdonnnj replied:
Appreciated!

mcdonnnj commented on molecule/default/tests/test_default.py:
```diff
@@ -12,7 +13,69 @@
+    # release is created the following line can be uncommented.
+    #
+    # See #3 for more details.
+    # assert s.exists, "systemd-resolved service does not exist."
```
If you want a test that is relatively equivalent I used (adapted for this project):

```python
load_state = s.systemd_properties["LoadState"]
assert load_state != "not-found", "systemd-resolved service does not exist."
```

jsf9k replied:
If the service does not exist then the other tests will fail, so I am fine with waiting for the fix to `testinfra`.  Can we close this conversation?

mcdonnnj reviewed (commented):
Looks pretty solid. I do have some feedback for your consideration.

jsf9k committed:
Separate tasks with a blank line

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Expand comment describing why we use follow=false

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Pull in the same version of ansible as is used in requirements-test.txt

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Uncomment Dependabot ignore directives from cisagov/skeleton-ansible-role

dav3r reviewed (approved):
👍 👍

mcdonnnj commented on meta/main.yml:
```diff
@@ -6,11 +6,13 @@
-    - skeleton
+    - resolved
+    - systemd
+    - systemdresolved
```
Just a thought (this should still comply with character restrictions).

```suggestion
    - systemd_resolved
```

jsf9k replied:
In fact this change does not comply with character restrictions:
```console
WARNING  Listing 1 violation(s) that are fatal
meta-no-tags: Tags must contain lowercase letters and digits only., invalid: 'systemd_resolved'
meta/main.yml:1

Read documentation for instructions on how to ignore specific rule violations.

             Rule Violation Summary              
 count tag          profile rule associated tags 
     1 meta-no-tags shared  metadata             

Failed: 1 failure(s), 0 warning(s) on 28 files. Last profile that met the validation criteria was 'safety'. Rating: 3/5 star
```

(I already tried it.)

mcdonnnj replied:
I stand corrected. Apologies I was going off of the [Galaxy documentation](https://docs.ansible.com/ansible/latest/dev_guide/collections_galaxy_meta.html) and didn't realize ansible-lint was doing it's own thang.

jsf9k replied:
Classic `ansible-lint`!

mcdonnnj commented on molecule/default/tests/test_default.py:
```diff
@@ -12,7 +13,69 @@
+    if host.system_info.distribution in ["amzn"]:
+        # /run/systemd/resolve/stub-resolv.conf is a symlink to
+        # /run/systemd/resolve/resolv.conf in AL2023, so the
+        # /etc/resolv.conf symlink resolves to the former.
```
Maybe it's just me but when I read former here it seems like it would be referring to `/run/systemd/resolve/stub-resolv.conf`.

```suggestion
        # /etc/resolv.conf symlink resolves to the latter.
```

mcdonnnj commented on molecule/default/tests/test_default.py:
```diff
@@ -12,7 +13,69 @@
+@pytest.mark.parametrize(
+    "dig_command",
+    [
+        "www.yahoo.com",
```
I still can't believe you used Yahoo! for this... (feel free to resolve this snarky feedback).

mcdonnnj reviewed (approved):
LGTM ✔ Two minor feedback items for your consideration.

jsf9k committed:
Correct comment

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>
